dcurve/src/internal/fetch.ts

```diff
@@ -0,0 +1,73 @@
+
+import { Contract } from 'ethers';
+import { Contribution, GrantRoundContributions, GrantRoundFetchArgs } from '../../src/types';
+import { abi as GRANT_ROUND_MANAGER_ABI } from '@dgrants/contracts/artifacts/contracts/GrantRoundManager.sol/GrantRoundManager.json';
+import { abi as GRANT_REGISTRY_ABI } from '@dgrants/contracts/artifacts/contracts/GrantRegistry.sol/GrantRegistry.json';
+
+const ERC20_ABI = ['function balanceOf(address) view returns (uint)'];
+
+/**
+ * Returns GrantRoundContributions describing the round and any contributions
+ * @param {GrantRoundFetchArgs}
+ */
+export const fetch = async (args: GrantRoundFetchArgs) => {
+  // fetch grantRoundManager
+  const roundManager = new Contract(args.grantRoundManager, GRANT_ROUND_MANAGER_ABI, args.provider);
+  // fetch the grantList
+  const registry = new Contract(args.grantRegistry, GRANT_REGISTRY_ABI, args.provider);
+
+  // set-up
+  const values = await Promise.all([
+    registry.getAllGrants(),
+    roundManager.donationToken(),
+    roundManager.queryFilter(roundManager.filters.GrantDonation()),
+  ]);
+
+  // collect the grants into a grantId->payoutAddress obj
+  const grantsList = values[0].reduce((grants: Record<string, string>, grant: Record<string, string>, key: number) => {
+    grants[key] = grant.payee;
+
+    return grants;
+  }, {});
+
+  // get donation token info
+  const donationToken = values[1];
+  const donationTokenContract = new Contract(donationToken, ERC20_ABI, args.provider);
+
+  // fetch & ignore Contributions
+  const grantDonations = values[2];
+  const contributions: Contribution[] = [];
+
+  // records matching contributions in the expected format (Contribution[])
+  grantDonations.forEach(async (contribution) => {
+    // get tx details to pull contributor details from
+    const tx = await contribution.getTransaction();
+
+    // check that the contribution is valid
+    const inRound = contribution?.args?.rounds.indexOf(args.grantRound) !== -1;
+    const isIgnoredGrant =
+      args.ignore && args.ignore.grants && args.ignore.grants?.indexOf(contribution?.args?.grantId.toNumber()) !== -1;
+    const isIgnoredContributor =
+      args.ignore && args.ignore.contributionAddress && args.ignore.contributionAddress?.indexOf(tx.from) !== -1;
+
+    // only include transactions from this grantRound which havent been ignored
+    if (inRound && !isIgnoredGrant && !isIgnoredContributor) {
+      contributions.push({
+        grantId: contribution?.args?.grantId.toNumber(),
+        amount: contribution?.args?.donationAmount / 10 ** args.supportedTokens[donationToken].decimals,
+        grantAddress: grantsList[contribution?.args?.grantId.toNumber()],
+        address: tx.from,
+      });
+    }
+  });
+
+  // return contributions as GrantRoundContributions
+  return {
+    grantRound: args.grantRound,
+    totalPot:
+      (await donationTokenContract.balanceOf(args.grantRound)) / 10 ** args.supportedTokens[donationToken].decimals,
+    currDecimals: 18,
```

Comment:
What's this? Unclear why it's hardcoded to 18

Reply:
Oh I see this used in the Merkle tree also, so it seems the idea is:
1. Get human readable amounts from this method
2. Pass them into the Merkle tree, along with `currDecimals`, to BigNumber-ify them

Hmm, maybe this method should take an optional payout token address and return it's decimal value here if provided and null or undefined otherwise? And maybe `matchTokenDecimals` would be more descriptive than `currDecimals`?

Reply:
> What's this? Unclear why it's hardcoded to 18

That should have been `args.supportedTokens[matchingToken].decimals` my bad!!

---

I think we will always want to avoid bigNumber for the clr calculation and only bring it back to generate the merkle - I think what you are suggesting here is similar to how its working, or am I missing something?

Reply:
> I think we will always want to avoid bigNumber for the clr calculation and only bring it back to generate the merkle - I think what you are suggesting here is similar to how its working, or am I missing something?

Agreed, yea l I think we are on the same page here! Sorry I may have been unclear/misunderstood previously